osu.Game.Rulesets.Taiko/Difficulty/Skills/Colour.cs

```diff
@@ -0,0 +1,119 @@
+        /// </summary>
+        private int currentMonoLength = 1;
+
+        /// <summary>
+        /// List of the last <see cref="mono_history_max_length"/> most recent mono patterns, with the most recent at the end of the list.
+        /// </summary>
+        private readonly List<int> monoHistory = new List<int>();
+
+        protected override double StrainValueOf(DifficultyHitObject current)
+        {
+            if (!(current.LastObject is Hit && current.BaseObject is Hit && current.DeltaTime < 1000))
+            {
+                previousHitType = null;
+                return 0.0;
+            }
+
+            var taikoCurrent = (TaikoDifficultyHitObject)current;
+
+            double objectStrain = 0.0;
+
+            if (taikoCurrent.HitType != null && previousHitType != null && taikoCurrent.HitType != previousHitType)
+            {
+                // The colour has changed.
+                objectStrain = 1.0;
+
+                if (monoHistory.Count < 2)
+                {
+                    // There needs to be at least two streaks to determine a strain.
+                    objectStrain = 0.0;
+                }
+                else if ((monoHistory[^1] + currentMonoLength) % 2 == 0)
+                {
+                    // The last streak in the history is guaranteed to be a different type to the current streak.
+                    // If the total number of notes in the two streaks is even, apply a penalty.
+                    objectStrain *= sameParityPenalty();
+                }
+
+                objectStrain *= repetitionPenalties();
+                currentMonoLength = 1;
+            }
+            else
+            {
+                currentMonoLength += 1;
+            }
+
+            previousHitType = taikoCurrent.HitType;
+            return objectStrain;
+        }
+
+        /// <summary>
+        /// The penalty to apply when the total number of notes in the two most recent colour streaks is even.
+        /// </summary>
+        private double sameParityPenalty() => 0.0;
+
+        /// <summary>
+        /// The penalty to apply due to the length of repetition in colour streaks.
+        /// </summary>
+        private double repetitionPenalties()
+        {
+            double penalty = 1.0;
+
+            monoHistory.Add(currentMonoLength);
+
+            if (monoHistory.Count > mono_history_max_length)
+                monoHistory.RemoveAt(0);
+
+            for (int l = 2; l <= mono_history_max_length / 2; l++)
+            {
+                for (int start = monoHistory.Count - l - 1; start >= 0; start--)
+                {
+                    bool samePattern = true;
+
+                    for (int i = 0; i < l; i++)
+                    {
+                        if (monoHistory[start + i] != monoHistory[monoHistory.Count - l + i])
+                        {
+                            samePattern = false;
+                        }
+                    }
```

Comment:
I'm really confused with this loop. Right now, as far as I can tell, it compares the following pairs of patterns:
```
(2, 3), (1, 3), (0, 3), (3, 4), (2, 4), (1, 4)
```
Is this correct? With how I understand what this was supposed to do `(0, 4)` isn't checked but should be, and `(3, 4)` is checked but it shouldn't be. I assume this meant to compare the newest 2 patterns against the rest (so patterns 0, 1, 2 against patterns 3, 4), but that requires a double loop as far as I know.

The rhythm repetition code is the same, so it presumably has the same issue.

Reply:
I believe this is working as intended. Intended behaviour is: The first time, it compares indexes
`2` with `3` and `3` with `4`
The next time, it compares indexes
`1` with `3` and `2` with `4`
The next time, it compares indexes
`0` with `3` and `1` with `4`